crypto/pkcs7/pk7_doit.c

```diff
@@ -480,14 +520,22 @@ BIO *PKCS7_dataDecode(PKCS7 *p7, EVP_PKEY *pkey, BIO *in_bio, X509 *pcert)
             }
 
             name = OBJ_nid2sn(OBJ_obj2nid(xa->algorithm));
+
+            (void)ERR_set_mark();
             evp_md = EVP_MD_fetch(p7_ctx->libctx, name, p7_ctx->propq);
-            if (evp_md == NULL) {
+            if (evp_md != NULL)
+                md = evp_md;
+            else
+                md = EVP_get_digestbyname(name);
+
+            if (md == NULL) {
                 PKCS7err(PKCS7_F_PKCS7_DATADECODE,
```

Comment:
You still have the missing ERR_clear_last_mark() here.

Reply:
Fixed. Thanks!